lib/tlIO/FFmpegWrite.cpp

```diff
@@ -350,40 +925,204 @@ namespace tl
 
             const auto timeRational = time::toRational(time.rate());
             p.avFrame->pts = av_rescale_q(
-                time.value(),
+                time.value() - p.videoStartTime.value(),
                 { timeRational.second, timeRational.first },
                 p.avVideoStream->time_base);
-            _encodeVideo(p.avFrame);
+
+            _encode(
+                    p.avCodecContext, p.avVideoStream, p.avFrame,
+                    p.avPacket);
         }
 
-        void Write::_encodeVideo(AVFrame* frame)
+        void Write::writeAudio(
+            const otime::TimeRange& inTimeRange,
+            const std::shared_ptr<audio::Audio>& audioIn,
+            const io::Options&)
         {
             TLRENDER_P();
+            
+            if (!audioIn || !p.avAudioFifo || audioIn->getSampleCount() == 0)
+                return;
+            
+            const auto& info = audioIn->getInfo();
+            if (!info.isValid())
+                return;
+            
+            int r = 0;
+            const auto timeRange = otime::TimeRange(
+                inTimeRange.start_time().rescaled_to(p.sampleRate),
+                inTimeRange.duration().rescaled_to(p.sampleRate));
+
+            int fifoSize = av_audio_fifo_size(p.avAudioFifo);
+                
+            if (timeRange.start_time().value() >=
+                p.totalSamples + p.audioStartSamples + fifoSize)
+            {
+                // If this is the start of the saving, store the start time.
+                if (p.totalSamples == 0)
+                {
+                    p.audioStartSamples = timeRange.start_time().value();
+                }
+
+                auto audioResampled = audioIn;
+                // Resample audio
+                if (p.resample)
+                {
+                    audioResampled = p.resample->process(audioIn);
+                }
+                
+                // Most codecs need non-interleaved audio.
+                std::shared_ptr<audio::Audio> audio;
+                if (p.avAudioPlanar)
+                    audio = planarDeinterleave(audioResampled);
+                else
+                    audio = audioResampled;
+                
+                uint8_t* data = audio->getData();
+                
+                // Allocate flatData pointers
+                const size_t channels = audio->getChannelCount();
+                const size_t stride = audio->getByteCount() / channels;
+                uint8_t** flatData = new uint8_t*[channels];
```

Comment:
How is this memory cleaned up?

Reply:
OOPS!!! I changed it to use a std::vector in private data.